File: algorithms/11_golang/techniques/backtracking.go
package main

import "fmt"

func subsets(nums []int) [][]int {
	res := [][]int{}
	cur := []int{}
	var dfs func(start int)
	dfs = func(start int) {
		cp := make([]int, len(cur))
		copy(cp, cur)
		res = append(res, cp)
		for i := start; i < len(nums); i++ {
			cur = append(cur, nums[i])
			dfs(i + 1)
			cur = cur[:len(cur)-1]
		}
	}
	dfs(0)
	return res
}

func permutations(nums []int) [][]int {
	res := [][]int{}
	cur := []int{}
	used := make([]bool, len(nums))
	var dfs func()
	dfs = func() {
		if len(cur) == len(nums) {
			cp := make([]int, len(cur))
			copy(cp, cur)
			res = append(res, cp)
			return
		}
		for i := 0; i < len(nums); i++ {
			if used[i] {
				continue
			}
			used[i] = true
			cur = append(cur, nums[i])
			dfs()
			cur = cur[:len(cur)-1]
			used[i] = false
		}
	}
	dfs()
	return res
}

func main() {
	s := subsets([]int{1, 2, 3})
	if len(s) != 8 {
		panic("backtracking: subsets count should be 8")
	}
	p := permutations([]int{1, 2, 3})
	if len(p) != 6 {
		panic("backtracking: permutations count should be 6")
	}
	fmt.Println("[Go Backtracking] Subsets (8) + permutations (6) verified.")
}

File: algorithms/06_javascript/techniques/backtracking.js
function subsets(nums) {
  const res = [];
  const cur = [];
  function dfs(i) {
    if (i === nums.length) { res.push([...cur]); return; }
    cur.push(nums[i]);
    dfs(i + 1);
    cur.pop();
    dfs(i + 1);
  }
  dfs(0);
  return res;
}

function permutations(nums) {
  const res = [];
  const used = new Array(nums.length).fill(false);
  const cur = [];
  function dfs() {
    if (cur.length === nums.length) { res.push([...cur]); return; }
    for (let i = 0; i < nums.length; i++) {
      if (used[i]) continue;
      used[i] = true;
      cur.push(nums[i]);
      dfs();
      cur.pop();
      used[i] = false;
    }
  }
  dfs();
  return res;
}

module.exports = { subsets, permutations };

if (require.main === module) {
  if (subsets([1, 2, 3]).length !== 8) throw new Error("subsets count failed");
  if (permutations([1, 2, 3]).length !== 6) throw new Error("permutations count failed");
  console.log("[JavaScript Backtracking] subsets & permutations verified");
}
